### PR TITLE
tracking attribute path into options[:attr_path]

### PR DESCRIPTION
**Pull request by calfzhou**
Make Grape::Entity keep tracking of current being exposed attribute's full path into `options[:attr_path]`.

### CODE REVIEW

**Comment by dblock** [comment]
I really like this, needs a CHANGELOG entry, please.

File: lib/grape_entity/entity.rb [hunk 1]
```diff
@@ -425,22 +425,27 @@ def serializable_hash(runtime_options = {})
...
-          elsif partial_output.is_a?(Array) && !partial_output.map { |o| o.respond_to?(:serializable_hash) }.include?(false)
-            partial_output.map(&:serializable_hash)
-          elsif partial_output.is_a?(Hash)
-            partial_output.each do |key, value|
-              partial_output[key] = value.serializable_hash if value.respond_to?(:serializable_hash)
+        parent_path = track_attr_path(attribute, opts)
+        begin
+          next unless should_return_attribute?(attribute, opts) && conditions_met?(exposure_options, opts)
+
+          partial_output = value_for(attribute, opts)
```
Comment (dblock) [inline lib/grape_entity/entity.rb]: Maybe construction of `partial_output` can be refactored into a method?

**Comment by calfzhou** [comment]
squashed and rebased to current master.

**Comment by dblock** [comment]
Merged.

File: .rubocop_todo.yml [hunk 2]
```diff
@@ -7,12 +7,12 @@
 
 # Offense count: 8
 Metrics/AbcSize:
-  Max: 51
+  Max: 53
```
Comment (marshall-lee) [inline .rubocop_todo.yml]: Metrics should not increase. Why we use RuboCop if we just stab it?

File: spec/grape_entity/entity_spec.rb [hunk 3]
```diff
@@ -991,6 +996,110 @@ def embedded
...
+            end
+          end
+
+          fresh_class.class_eval do
+            expose :characteristics, using: EntitySpec::CharacterEntity
+
+            protected
+
+            def self.path_for(attribute)
+              attribute == :characteristics ? :character : super
```
Comment (marshall-lee) [inline spec/grape_entity/entity_spec.rb]: I don't like this idea. This thing should be implemented in exposure parameters. For example:

```ruby
expose :characteristics, using: EntitySpec::CharacterEntity, attr_path: proc { :character }
```
What if we have more attributes which path we need to override? We'll end up with conditional hell in `path_for`.

**Comment by marshall-lee** [comment]
@calfzhou, @dblock 

I backported this feature while working on rebasing #151. But I changed your specs because in the new design there's no place for methods like `key_for`, `name_for` and (newly introduced) `path_for`.

Instead of this I implemented a `:attr_path` exposure option that adds the same possibility:
```ruby
expose :charactersistics, attr_path: -> (obj, opts) { :character }
expose :need_to_skip, attr_path: -> (obj, opts) { nil }
```

Maybe it would be useful to also add these possibilities in feature:
```ruby
expose :need_to_skip, attr_path: :character # just constant
expose :need_to_skip, attr_path: nil # just skip
```

File: .rubocop_todo.yml (hunk 2 above)
Comment (dblock) [inline .rubocop_todo.yml]: Personally I don't care as much, I think it's a good flag and if it makes sense, I would not make it a barrier.

File: spec/grape_entity/entity_spec.rb (hunk 3 above)
Comment (dblock) [inline spec/grape_entity/entity_spec.rb]: :+1:  on what @marshall-lee says, would love a refactor @calfzhou

**Comment by dblock** [comment]
Thanks @marshall-lee your changes make sense to me.

**Comment by calfzhou** [comment]
@marshall-lee 's `attr_path` exposure is pretty good!

File: .rubocop_todo.yml (hunk 2 above)
Comment (marshall-lee) [inline .rubocop_todo.yml]: Yes I think the same but I just want to see things *before* changing `.rubocop_todo.yml`. There could be something we can suggest to easily fix and if there is not then we suggest to just regenerate `.rubocop_todo.yml` :)

Comment (marshall-lee) [inline .rubocop_todo.yml]: As features and their correctness are superior to style pedantry, I want to push #160.

I think the reason that people just stab RuboCop offenses is that it runs *before* RSpec. It prevents maintainers to just see the feature implemented correctly or not.

Comment (dblock) [inline .rubocop_todo.yml]: I would like to disable this offense in Rubocop, and be able to run `--auto-gen-config` for everything else. However when you try to do the latter it ignores what's in `.rubocop.yml`, which defeats the whole purpose.

Feel free to switch it to run after, btw.

**Pull request merged**